functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,179 @@
+package transformer
+
+import (
+	"fmt"
+
+	"github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/set-image/custom"
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"sigs.k8s.io/kustomize/api/image"
+	"sigs.k8s.io/kustomize/api/types"
+)
+
+// SetImage supports the set-image workflow, it uses Config to parse functionConfig, Transform to change the image
+type SetImage struct {
+	// Image is the desired image
+	Image types.Image `json:"image,omitempty" yaml:"image,omitempty"`
+	// ConfigMap keeps the data field that holds image information
+	ConfigMap map[string]string `json:"data,omitempty" yaml:"data,omitempty"`
+	// ONLY for kustomize, AdditionalImageFields is the user supplied fieldspec
+	AdditionalImageFields types.FsSlice `json:"additionalImageFields,omitempty" yaml:"additionalImageFields,omitempty"`
+	// context logs each detailed result
+	context *fn.Context
+	// resultCount logs the total count image change
+	resultCount int
+}
+
+// Config transforms the data from ConfigMap to SetImage struct
+func (imageTrans *SetImage) Config() error {
+	for key, val := range imageTrans.ConfigMap {
+		if key == "name" {
+			imageTrans.Image.Name = val
+		} else if key == "newName" {
+			imageTrans.Image.NewName = val
+		} else if key == "newTag" {
+			imageTrans.Image.NewTag = val
+		} else if key == "digest" {
+			imageTrans.Image.Digest = val
+		} else {
+			return fmt.Errorf("SubObject has unmatched field type: `data`")
+		}
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (imageTrans *SetImage) validateInput() error {
+	if imageTrans.Image.Name == "" {
+		return fmt.Errorf("missing image name or typo")
+	}
+	if imageTrans.Image.NewName == "" && imageTrans.Image.NewTag == "" && imageTrans.Image.Digest == "" {
+		return fmt.Errorf("missing image newName, newTag, or digest, could be typo or missing field")
+	}
+	if imageTrans.Image.NewTag != "" && imageTrans.Image.Digest != "" {
+		return fmt.Errorf("image newTag and digest both set")
+	}
+	return nil
+}
+
+func (imageTrans *SetImage) setPodSpecContainers(o *fn.KubeObject) error {
+	if spec := o.GetMap("spec"); spec != nil {
+		if template := spec.GetMap("template"); template != nil {
+			if podSpec := template.GetMap("spec"); podSpec != nil {
+				for _, vecObj := range podSpec.GetSlice("containers") {
+					if err := imageTrans.updateImages(vecObj, &imageTrans.Image, o); err != nil {
+						return err
+					}
+				}
+				for _, vecObj := range podSpec.GetSlice("iniContainers") {
+					if err := imageTrans.updateImages(vecObj, &imageTrans.Image, o); err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+	return nil
```

Comment:
My previous comment seems to be not resolved  https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/904#discussion_r946246928

Reply:
ditto for the following `setPodContainers`, `hasPodSpecContainers` functions.